backends/simplego/dotgeneral_fastpath.go

```diff
@@ -0,0 +1,201 @@
+package simplego
+
+import (
+	"github.com/gomlx/go-xla/pkg/types/dtypes"
+	"github.com/gomlx/gomlx/pkg/core/shapes"
+)
+
+// isStandardMatmul checks if the DotGeneral operation is a standard matrix multiplication
+// that doesn't require any transposition or complex axis manipulation.
+//
+// Standard patterns that can skip normalization:
+// 1. Matrix × Matrix: [M, K] × [K, N] → [M, N] (contracting on last axis of lhs, first of rhs)
+// 2. Matrix × Vector: [M, K] × [K] → [M] (contracting on last axis of lhs, only axis of rhs)
+// 3. Batched MatMul: [B, M, K] × [B, K, N] → [B, M, N] (batch on first axis)
+//
+// Returns true if we can use the fast path (no transpose needed).
+func isStandardMatmul(lhsShape, rhsShape shapes.Shape, lhsContractingAxes, rhsContractingAxes, lhsBatchAxes, rhsBatchAxes []int) bool {
+	lhsRank := lhsShape.Rank()
+	rhsRank := rhsShape.Rank()
+
+	// Check for standard matrix-matrix multiplication: [M, K] × [K, N]
+	if lhsRank == 2 && rhsRank == 2 &&
+		len(lhsContractingAxes) == 1 && len(rhsContractingAxes) == 1 &&
+		len(lhsBatchAxes) == 0 && len(rhsBatchAxes) == 0 {
+		// Contracting: lhs last axis (1) with rhs first axis (0)
+		if lhsContractingAxes[0] == 1 && rhsContractingAxes[0] == 0 {
+			return true
+		}
+	}
+
+	// Check for matrix-vector multiplication: [M, K] × [K]
+	if lhsRank == 2 && rhsRank == 1 &&
+		len(lhsContractingAxes) == 1 && len(rhsContractingAxes) == 1 &&
+		len(lhsBatchAxes) == 0 && len(rhsBatchAxes) == 0 {
+		// Contracting: lhs last axis (1) with rhs only axis (0)
+		if lhsContractingAxes[0] == 1 && rhsContractingAxes[0] == 0 {
+			return true
+		}
+	}
+
+	// Check for batched matrix multiplication: [B, M, K] × [B, K, N]
+	if lhsRank == 3 && rhsRank == 3 &&
+		len(lhsContractingAxes) == 1 && len(rhsContractingAxes) == 1 &&
+		len(lhsBatchAxes) == 1 && len(rhsBatchAxes) == 1 {
+		// Batch on first axis (0)
+		if lhsBatchAxes[0] == 0 && rhsBatchAxes[0] == 0 {
+			// Contracting: lhs axis 2 with rhs axis 1
+			if lhsContractingAxes[0] == 2 && rhsContractingAxes[0] == 1 {
+				return true
+			}
+		}
+	}
+
+	// Check for multi-batch matmul: [B1, B2, M, K] × [B1, B2, K, N]
+	if lhsRank == 4 && rhsRank == 4 &&
+		len(lhsContractingAxes) == 1 && len(rhsContractingAxes) == 1 &&
+		len(lhsBatchAxes) == 2 && len(rhsBatchAxes) == 2 {
+		// Batch on first two axes
+		if lhsBatchAxes[0] == 0 && lhsBatchAxes[1] == 1 &&
+			rhsBatchAxes[0] == 0 && rhsBatchAxes[1] == 1 {
+			// Contracting: lhs axis 3 with rhs axis 2
+			if lhsContractingAxes[0] == 3 && rhsContractingAxes[0] == 2 {
+				return true
+			}
+		}
+	}
+
+	return false
+}
+
+// isMemoryContiguous checks if the tensor layout is already contiguous in memory
```

Comment:
I think the name here should be more "isBatchCrossContract", no ? Since the memory is always contiguous.